### PR TITLE
Create a way to download themes from a github repository

### PR DESCRIPTION
**Pull request by YodaForce157**
TODO:

- [x] Remove " from label
- [x] Create download button
- [x] Fix positioning
- [x] Create preview button that either:
- A: ~~Links the user to an image~~
- B: Shows the image in Railroad
- C: ~~Applies a small sample of the theme to another window with example text/buttons etc~~
- [x] Create a way to recognise if a theme is already downloaded or not

### CODE REVIEW

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadCell.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        themeDownloadNode.setAlignment(Pos.CENTER_LEFT);
+
+        downloadButton.setOnAction(action -> {
+            try {
+                boolean res = ThemeDownloadManager.downloadTheme(new URL(themeDownloadNode.jsonProperty().get().get("download_url").getAsString()));
+                this.downloadButton.setKey(res ? "railroad.home.settings.appearance.installed" : "railroad.home.settings.appearance.download");
+                this.downloadButton.setDisable(res);
+            } catch (MalformedURLException e) {
+                throw new RuntimeException(e);
+            }
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadCell.java]: Replace with logger

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+            LOGGER.info("Downloading theme: {}", fileName);
+            try {
+                Thread downloadThread = new Thread(() -> FileHandler.copyUrlToFile(url.toString(), Paths.get(getThemesDir().toString(), fileName)));
+                downloadThread.start();
+                downloadThread.join();
+
+                LOGGER.info("Completed theme download");
+            } catch (Exception e) {
+                throw new RuntimeException("Exception downloading theme", e);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java]: Replace with logger

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    public static ObservableList<String> getDownloaded() {
+        Stream<String> themes;
+        Path dir = getThemesDir();
+
+        if(Files.notExists(dir)) {
+            try {
+                Files.createDirectory(dir);
+            } catch (IOException e) {
+                throw new RuntimeException("Could not create themes directory", e);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java]: Replace with logger

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            }
+        }
+
+        try {
+            themes = Files.list(dir).map(e -> e.getFileName().toString());
+        } catch (IOException e) {
+            throw new RuntimeException("Could not fetch installed themes", e);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java]: Replace with logger

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            LOGGER.info("FETCHING THEMES");
+            URL conUrl = new URL(url);
+            HttpURLConnection connection = (HttpURLConnection) conUrl.openConnection();
+            connection.setRequestMethod("GET");
+            connection.connect();
+
+            int resCode = connection.getResponseCode();
+
+            if(resCode != 200) {
+                throw new RuntimeException("THEME DOWNLOADER ERROR " + resCode);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java]: Replace with logger

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                inline += scanner.nextLine();
+            }
+
+            scanner.close();
+            connection.disconnect();
+
+            jsonRes = JsonParser.parseString(inline).getAsJsonArray();
+            lastRefreshed = ZonedDateTime.now();
+        } catch (IOException e) {
+            throw new RuntimeException("Could not list themes from github.");
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadManager.java]: Replace with logger

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadPane.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        var title = new LocalizedLabel("railroad.home.settings.appearance.downloadtheme");
+        title.setAlignment(Pos.TOP_CENTER);
+        title.setStyle("-fx-font-weight: bold; -fx-font-size: 20; -fx-alignment: center");
+
+        listView.getItems().addAll(observableThemes);
+
+        pane.getChildren().addAll(title, listView);
+        pane.setAlignment(Pos.TOP_CENTER);
+
+        Scene scene = new Scene(pane, 450, 450);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeDownloadPane.java]:
```suggestion
        var scene = new Scene(pane, 450, 450);
```

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeExamplePane.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import javafx.stage.Stage;
+
+import static io.github.railroad.Railroad.LOGGER;
+
+public class ThemeExamplePane {
+    public ThemeExamplePane(final String theme) {
+        var stage = new Stage();
+        var pane = new RRVBox();
+
+        Image image = new Image("https://github.com/Railroad-Team/Themes/blob/main/" + theme.replace(".css","") + ".png?raw=true");
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeExamplePane.java]:
```suggestion
        var image = new Image("https://github.com/Railroad-Team/Themes/blob/main/" + theme.replace(".css","") + ".png?raw=true");
```

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeExamplePane.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        LOGGER.debug("getting image {}", image.getUrl());
+        ImageView view = new ImageView(image);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeExamplePane.java]:
```suggestion
        var view = new ImageView(image);
```

File: src/main/java/io/github/railroad/settings/ui/themes/ThemeExamplePane.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+        view.fitHeightProperty().bind(stage.heightProperty());
+        view.fitWidthProperty().bind(stage.widthProperty());
+        view.setPreserveRatio(true);
+
+        pane.getChildren().add(view);
+
+        Scene scene = new Scene(pane, 500, 350);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/settings/ui/themes/ThemeExamplePane.java]:
```suggestion
        var scene = new Scene(pane, 500, 350);
```

File: src/main/resources/io/github/railroad/lang/es_es.lang [hunk 11]
```diff
@@ -31,6 +31,11 @@ railroad.home.settings.general.github.disconnect=Desconectar
 railroad.home.settings.general.language=Idioma:
 
 railroad.home.settings.appearance.theme=Seleccionar un tema
+railroad.home.settings.appearance.downloadtheme=Download themesE
+railroad.home.settings.appearance.notfound=Themes not foundE
+railroad.home.settings.appearance.download=DownloadE
+railroad.home.settings.appearance.installed=InstalledE
+railroad.home.settings.appearance.preview=PreviewE
```
Comment (DaRealTurtyWurty) [inline src/main/resources/io/github/railroad/lang/es_es.lang]: What's going on here?

**Review by DaRealTurtyWurty — CHANGES_REQUESTED** [review]

File: src/main/resources/io/github/railroad/lang/es_es.lang (hunk 11 above)
Comment (YodaForce157) [reply src/main/resources/io/github/railroad/lang/es_es.lang]: Oops that was just to check it worked properly 😂 
Will translate it now.

**Review by DaRealTurtyWurty — APPROVED** [review]

**Pull request merged**